010.md

```diff
@@ -0,0 +1,451 @@
+
+For **disconnected messages** we try to place them in the correct position by
+timestamp alone. It could be a good idea to indicate in UI that this message may
+be in the wrong place and is missing context.
+
+
+```mermaid
+flowchart RL
+A(A):::root;B(B);C(C);D(D);M(M);N(N);Y(Y):::disconnected
+N-->Y-->M-->D-->C-->B-->A
+classDef default fill: #e953da, stroke-width: 0, color: white
+classDef root fill: #4c44cf, stroke-width: 0, color: white
+classDef disconnected stroke:none,fill:#e95353,color:none;
+```
+_Figure. showing a linearised thread tangle, along with a "best guess" placement
+of disconnected message `Y`._
+
+
+### 5.2. Scuttlebutt Markdown
+
+The `text` field of `post` messages is Scuttlebutt flavoured Markdown, which is
+basic Markdown with a couple of important privacy enhancing features.
+
+#### 5.2.1 Image rendering
+
+DO NOT render images that link to the web. This leaks data, which allows
+tracking of users.
+```markdown
+![unsafe image](https://upload.wikimedia.org/wikipedia/commons/1/11/Panopticon.jpg)
+```
+Such images MUST instead be rendered as a link with an alert/ warning
+(see 6.2.2).
+
+ONLY render images that link to blobs that are hosted locally:
+```markdown
+![snails](&7AUsuVcxUzIJZPBBLqR3O2KvIiuECgdYq9lzEh0jcS8=.sha256)
+```
+
+#### 5.2.2 Link rendering
+
+When a link to the web is present, the interface MUST render an alert to users
+on-click. Similar to 6.2.1, this is to prevent accidental tracking.
+
+```markdown
+[click here](www.google.com)
+```
+
+#### 5.2.3 Old @mentions (deprecated)
+
+Older messages used `mentions` in a fancy way so that the text didn't have
+honking ugly markdown links to feedIds in it, but most clients have given
+this up
+
+```javascript!
+{
+  type: 'post'
+  text: 'This was an innovation @pfraze came up with',
+  mentions: [
+    {
+      link: '@hxGxqPrplLjRG2vtjQL87abX4QKqeLgCwQpS730nNwE=.ed25519',
+      name: '@pfraze'
+    }
+  ]
+}
+```
+The mentions are then used to map the `text` to this before rendering:
+```markdown
+This was an innovation [@pfraze](@hxGxqPrplLjRG2vtjQL87abX4QKqeLgCwQpS730nNwE=.ed25519)
+came up with',
+```
+
+### 5.3. Blocked authors
+
+If some message `B` is from a blocked author, we may still use its tangle data
+for causal ordering, but not render that message in the thread. 
+  
+The message MUST NOT be displayed to by deault in the UI, but it is up to the
+particular client whether this means
+- not rendering it at all
+- rendering an empty skeleton message saying "message from blocked author"
+- rendering a skeleton saying "message from blocked author, do you want to
+  display?"
+  
+
+### 5.4. Content warnings
+
+If the post contains a `contentWarning` field, the UI SHOULD NOT display the
+raw content and instead display a message which:
+1. alerts that this message has a content warning
+2. displays the String in `contentWarning`
+
+If the user clicks "show content" (or similar) then the message should be
+rendered as usual.
+
+
```

Comment:
```suggestion
## Reference implementation

- [ssb-threads](https://github.com/ssbc/ssb-threads)

```

Reply:
Need to check with you that you actually do the ordering like this spec suggests
I've seen the linearizing done a couple ways since I've worked on it

I think doing a sort like if A is upstream (in previous chain) of B then -1, if B is upstream of A, then +1, otherwise sort by timestamp?

Reply:
ssb-threads uses [ssb-sort](https://github.com/ssbc/ssb-sort/) (which is the reference implementation for the causal/topo sort) so I think it's correct? https://github.com/ssbc/ssb-threads/blob/ca06fcafabf1870bcb29ea24d899235249b7966f/src/index.ts#L279

Reply:
cool. Yeah I remember the logic now, I think it's what I said...
https://github.com/ssbc/ssb-sort/blob/master/index.js#L32-L49